certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+//// ## Verification of Splits.sol
+//// 
+//// `Splits.sol` is part of the Drips protocol
+//// 
+//// The splits functionality, as explained in DripsHub.sol by the developers:
+//// The user can share collected funds with other users by using splits.
+//// When collecting, the user gives each of their splits receivers a fraction of the received funds.
+//// Funds received from splits are available for collection immediately regardless of the cycle.
+//// They aren't exempt from being split, so they too can be split when collected.
+//// Users can build chains and networks of splits between each other.
+//// Anybody can request collection of funds for any user,
+//// which can be used to enforce the flow of funds in the network of splits.
+//// 
+//// ### Assumptions and Simplifications
+//// - All the internal methods are wrapped in CVT callable versions
+//// - Dummy functions used to control the SplitsReceiver[] receivers
+//// - When testing the split weight functionality we safely assumed weight < _TOTAL_SPLITS_WEIGHT
+////   since the method _assertSplitsValid() validates them before assignment
+//// 
+//// ### Properties
+
+
+/// @notice Calculate results of splitting an amount using the current splits configuration.
+/// We verify that amount == collectableAmt + splitAmt, including that no overflow occurs
+rule correctnessOfSplitResults() {
+    env e; uint256 userId; uint128 amount; uint128 collectableAmt; uint128 splitAmt;
+
+    collectableAmt, splitAmt = splitResults(e, userId, true, amount);
+
+    assert amount >= collectableAmt;
+    assert amount >= splitAmt;
+    assert amount == collectableAmt + splitAmt;
+}
+
+
+/// @notice Splits user's received but not split yet funds among receivers.
+/// @param collectableAmt The amount made collectable for the user
+/// @param splitAmt The amount split to the user's splits receivers
+rule correctnessOfSplit() {
+    env e; uint256 userId; uint256 assetId; uint128 collectableAmt; uint128 splitAmt;
+
+    uint128 splittableBefore;   uint128 collectableBefore;
+    uint128 splittableAfter;    uint128 collectableAfter;
+    
+    splittableBefore = splittable(e, userId, assetId);
+    collectableBefore = collectable(e, userId, assetId);
+
+    collectableAmt, splitAmt = split(e, userId, assetId, true);
```

Comment:
If you get the collectable and splitable amounts from the function, can't you check more precisely the relation between the old and new values, i.e. that the difference equals the amount?

Reply:
In the last assert, the difference might not equal exactly in the case that the splitter set himself as a splitReceiver.

Or maybe you meant that it should be:
`assert collectableBefore + collectableAmt == collectableAfter`
which I will update

Reply:
I think it's worth checking the case in which a splitter set himself as the receiver.

Reply:
I agree and later on I have a more elaborate rule where the splitReceivers are controlled. This rule was supposed to be a short and general one without having to check anything about the splitReceivers.